File: asymmetric_crypto.py
import base64
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding, rsa


def generate_key_pair():
    private_key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
    public_key = private_key.public_key()

    public_bytes = public_key.public_bytes(
        encoding=serialization.Encoding.DER,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )

    private_bytes = private_key.private_bytes(
        encoding=serialization.Encoding.DER,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    )

    public_key_b64 = base64.b64encode(public_bytes).decode("utf-8")
    private_key_b64 = base64.b64encode(private_bytes).decode("utf-8")

    return public_key, private_key, public_key_b64, private_key_b64


def encrypt_message(public_key, message: str) -> bytes:
    plaintext = message.encode("utf-8")
    return public_key.encrypt(
        plaintext,
        padding.PKCS1v15(),
    )


def load_private_key(private_key_b64: str):
    der = base64.b64decode(private_key_b64)
    return serialization.load_der_private_key(der, password=None)


def main():
    print("=================================")
    print("   RSA Asymmetric Cryptography")
    print("=================================\n")

    message = input("Enter message to encrypt: ")

    print("\nGenerating RSA Keys...")
    public_key, private_key, public_key_b64, private_key_b64 = generate_key_pair()
    print("Keys Generated Successfully!\n")

    print("---------- PUBLIC KEY ----------")
    print(public_key_b64)

    print("\n---------- PRIVATE KEY ----------")
    print(private_key_b64)

    encrypted_bytes = encrypt_message(public_key, message)
    encrypted_b64 = base64.b64encode(encrypted_bytes).decode("utf-8")

    print("\nEncrypted Message:")
    print(encrypted_b64)

    print("\nPaste the Private Key for Decryption:")
    entered_key = input().strip()

    loaded_private_key = load_private_key(entered_key)
    decrypted_bytes = loaded_private_key.decrypt(
        base64.b64decode(encrypted_b64),
        padding.PKCS1v15(),
    )
    decrypted_text = decrypted_bytes.decode("utf-8")

    print("\nDecrypted Message:")
    print(decrypted_text)


if __name__ == "__main__":
    main()

File: AsymmetricCrypto.java
import java.security.KeyPairGenerator;
import java.security.KeyPair;
import java.security.PublicKey;
import java.security.PrivateKey;
import java.security.KeyFactory;
import java.security.spec.PKCS8EncodedKeySpec;
import javax.crypto.Cipher;
import java.util.Base64;
import java.util.Scanner;

public class AsymmetricCrypto {

    // Encryption Method
    private static byte[] encrypt(byte[] data, PublicKey key) throws Exception {
        Cipher cipher = Cipher.getInstance("RSA");
        cipher.init(Cipher.ENCRYPT_MODE, key);
        return cipher.doFinal(data);
    }

    // Decryption Method
    private static byte[] decrypt(byte[] data, PrivateKey key) throws Exception {
        Cipher cipher = Cipher.getInstance("RSA");
        cipher.init(Cipher.DECRYPT_MODE, key);
        return cipher.doFinal(data);
    }

    // Convert Base64 String into PrivateKey
    private static PrivateKey getPrivateKey(String key) throws Exception {
        byte[] keyBytes = Base64.getDecoder().decode(key);

        PKCS8EncodedKeySpec spec =
                new PKCS8EncodedKeySpec(keyBytes);

        KeyFactory factory = KeyFactory.getInstance("RSA");

        return factory.generatePrivate(spec);
    }

    public static void main(String[] args) throws Exception {

        Scanner sc = new Scanner(System.in);

        System.out.println("=================================");
        System.out.println("   RSA Asymmetric Cryptography");
        System.out.println("=================================\n");

        // Input message
        System.out.print("Enter message to encrypt: ");
        String message = sc.nextLine();

        // Generate RSA Key Pair
        System.out.println("\nGenerating RSA Keys...");

        KeyPairGenerator keyGen =
                KeyPairGenerator.getInstance("RSA");

        keyGen.initialize(2048);

        KeyPair pair = keyGen.generateKeyPair();

        PublicKey publicKey = pair.getPublic();
        PrivateKey privateKey = pair.getPrivate();

        System.out.println("Keys Generated Successfully!\n");

        // Display Public Key
        System.out.println("---------- PUBLIC KEY ----------");
        System.out.println(
                Base64.getEncoder()
                        .encodeToString(publicKey.getEncoded())
        );

        // Display Private Key
        System.out.println("\n---------- PRIVATE KEY ----------");
        String privateKeyText =
                Base64.getEncoder()
                        .encodeToString(privateKey.getEncoded());

        System.out.println(privateKeyText);

        // Encrypt Message
        byte[] encryptedBytes =
                encrypt(message.getBytes(), publicKey);

        String encryptedText =
                Base64.getEncoder()
                        .encodeToString(encryptedBytes);

        System.out.println("\nEncrypted Message:");
        System.out.println(encryptedText);

        // Manual Private Key Input
        System.out.println("\nPaste the Private Key for Decryption:");
        String enteredKey = sc.nextLine();

        // Convert entered key to PrivateKey object
        PrivateKey enteredPrivateKey =
                getPrivateKey(enteredKey);

        // Decrypt
        byte[] decryptedBytes =
                decrypt(encryptedBytes, enteredPrivateKey);

        String decryptedText =
                new String(decryptedBytes);

        System.out.println("\nDecrypted Message:");
        System.out.println(decryptedText);

        sc.close();
    }
}